### PR TITLE
[libc] Disable hidden visibility for LIBC_NAMESPACE with GCC

### PR DESCRIPTION
**Pull request by petrhosek**
GCC emits a warning when using the visibility attribute which needs to be diagnosed and addressed, but this change should unbreak the GCC build as a temporary workaround.

The issue is tracked as #98548.

### CODE REVIEW

File: libc/src/__support/macros/config.h [hunk 1]
```diff
@@ -28,6 +28,7 @@
 #define LIBC_HAS_FEATURE(f) 0
 #endif
 
+#ifndef __GNUC__
```
Comment (lntue) [inline libc/src/__support/macros/config.h:31]: I think clang also define this macro.  Probably `#ifdef __clang__` is safer?

Comment (michaelrj-google) [inline libc/src/__support/macros/config.h:31]: doesn't clang define `__GNUC__` for compatibility? Would it be better to instead use `#ifdef __clang__` instead?

Comment (frobtech) [inline libc/src/__support/macros/config.h:31]: `__GNUC__` is defined by many compilers, including Clang.
You need `#ifdef __clang__` here.

**Comment by jhuber6** [comment]
What is the warning? This should work in C++17 as far as I know https://godbolt.org/z/j94hxsTGa.

File: libc/src/__support/macros/config.h (hunk 1 above)
Comment (petrhosek) [reply libc/src/__support/macros/config.h]: Done.

Comment (petrhosek) [reply libc/src/__support/macros/config.h]: Done.

Comment (petrhosek) [reply libc/src/__support/macros/config.h]: Done.

**Comment by michaelrj-google** [comment]
> What is the warning? This should work in C++17 as far as I know https://godbolt.org/z/j94hxsTGa.

From the buildbot:

```
/home/llvm-libc-buildbot/buildbot-worker/libc-x86_64-debian-fullbuild/libc-x86_64-debian-gcc-fullbuild-dbg/llvm-project/libc/src/__support/big_int.h: In instantiation of ‘constexpr __llvm_libc_19_0_0_git::cpp::array<word, N> __llvm_libc_19_0_0_git::multiword::shift(__llvm_libc_19_0_0_git::cpp::array<word, N>, size_t) [with Direction direction = __llvm_libc_19_0_0_git::multiword::RIGHT; bool is_signed = false; word = long unsigned int; long unsigned int N = 5; size_t = long unsigned int]’:
/home/llvm-libc-buildbot/buildbot-worker/libc-x86_64-debian-fullbuild/libc-x86_64-debian-gcc-fullbuild-dbg/llvm-project/libc/src/__support/big_int.h:760:61:   required from ‘constexpr __llvm_libc_19_0_0_git::BigInt<Bits, Signed, WordType> __llvm_libc_19_0_0_git::BigInt<Bits, Signed, WordType>::operator>>(size_t) const [with long unsigned int Bits = 320; bool Signed = false; WordType = long unsigned int; size_t = long unsigned int]’
/home/llvm-libc-buildbot/buildbot-worker/libc-x86_64-debian-fullbuild/libc-x86_64-debian-gcc-fullbuild-dbg/llvm-project/libc/src/__support/float_to_string.h:390:29:   required from here
/home/llvm-libc-buildbot/buildbot-worker/libc-x86_64-debian-fullbuild/libc-x86_64-debian-gcc-fullbuild-dbg/llvm-project/libc/src/__support/big_int.h:281:28: error: ‘__llvm_libc_19_0_0_git::multiword::shift<__llvm_libc_19_0_0_git::multiword::RIGHT, false, long unsigned int, 5>(__llvm_libc_19_0_0_git::cpp::array<long unsigned int, 5>, size_t)::<lambda(size_t)>’ declared with greater visibility than the type of its field ‘__llvm_libc_19_0_0_git::multiword::shift<__llvm_libc_19_0_0_git::multiword::RIGHT, false, long unsigned int, 5>(__llvm_libc_19_0_0_git::cpp::array<long unsigned int, 5>, size_t)::<lambda(size_t)>::<array capture>’ [-Werror=attributes]
  281 |   const auto safe_get_at = [&](size_t index) -> word {
      |                            ^~~~~~~~~~~~~~~~~~~~~~~~~~~
  282 |     // return appropriate value when accessing out of bound elements.
      |     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  283 |     const int i = at(index);
      |     ~~~~~~~~~~~~~~~~~~~~~~~~
  284 |     if (i < 0)
      |     ~~~~~~~~~~              
  285 |       return 0;
      |       ~~~~~~~~~             
  286 |     if (i >= int(N))
      |     ~~~~~~~~~~~~~~~~        
  287 |       return is_neg ? -1 : 0;
      |       ~~~~~~~~~~~~~~~~~~~~~~~
  288 |     return array[i];
      |     ~~~~~~~~~~~~~~~~        
  289 |   };
      |   ~                         
```

**Review by michaelrj-google — APPROVED** [review]

**Pull request merged**